File: apps/demo-be-csas/src/DemoBeCsas/Auth/JwtService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.IdentityModel.Tokens;

namespace DemoBeCsas.Auth;

public interface IJwtService
{
    string CreateAccessToken(string userId, string username, string role);
    string CreateRefreshToken(string userId);
    ClaimsPrincipal? DecodeToken(string token);
    string? GetJti(string token);
    DateTimeOffset? GetIssuedAt(string token);
}

public class JwtService(IConfiguration config) : IJwtService
{
    private readonly string _secret =
        config["APP_JWT_SECRET"]
        ?? throw new InvalidOperationException("APP_JWT_SECRET not configured");

    private SymmetricSecurityKey Key =>
        new(Encoding.UTF8.GetBytes(_secret));

    public string CreateAccessToken(string userId, string username, string role)
    {
        var creds = new SigningCredentials(Key, SecurityAlgorithms.HmacSha256);
        var jti = Guid.NewGuid().ToString();
        var now = DateTimeOffset.UtcNow;

        var token = new JwtSecurityToken(
            claims:
            [
                new Claim(JwtRegisteredClaimNames.Sub, userId),
                new Claim("username", username),
                new Claim("role", role),
                new Claim(JwtRegisteredClaimNames.Jti, jti),
                new Claim(JwtRegisteredClaimNames.Iat, now.ToUnixTimeSeconds().ToString(), ClaimValueTypes.Integer64),
            ],
            expires: DateTime.UtcNow.AddMinutes(15),
            signingCredentials: creds
        );

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    public string CreateRefreshToken(string userId)
    {
        var creds = new SigningCredentials(Key, SecurityAlgorithms.HmacSha256);
        var jti = Guid.NewGuid().ToString();
        var now = DateTimeOffset.UtcNow;

        var token = new JwtSecurityToken(
            claims:
            [
                new Claim(JwtRegisteredClaimNames.Sub, userId),
                new Claim("token_type", "refresh"),
                new Claim(JwtRegisteredClaimNames.Jti, jti),
                new Claim(JwtRegisteredClaimNames.Iat, now.ToUnixTimeSeconds().ToString(), ClaimValueTypes.Integer64),
            ],
            expires: DateTime.UtcNow.AddDays(7),
            signingCredentials: creds
        );

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    public ClaimsPrincipal? DecodeToken(string token)
    {
        // Disable inbound claim type mapping to preserve original JWT claim names (e.g. "sub" stays "sub")
        var handler = new JwtSecurityTokenHandler { MapInboundClaims = false };
        var validationParams = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = Key,
            ValidateIssuer = false,
            ValidateAudience = false,
            ClockSkew = TimeSpan.Zero,
        };

        try
        {
            return handler.ValidateToken(token, validationParams, out _);
        }
        catch
        {
            return null;
        }
    }

    public string? GetJti(string token)
    {
        try
        {
            var handler = new JwtSecurityTokenHandler();
            var jwt = handler.ReadJwtToken(token);
            return jwt.Id;
        }
        catch
        {
            return null;
        }
    }

    public DateTimeOffset? GetIssuedAt(string token)
    {
        try
        {
            var handler = new JwtSecurityTokenHandler();
            var jwt = handler.ReadJwtToken(token);
            return jwt.IssuedAt == default
                ? null
                : new DateTimeOffset(jwt.IssuedAt, TimeSpan.Zero);
        }
        catch
        {
            return null;
        }
    }
}

File: apps/demo-be-fsharp-giraffe/src/DemoBeFsgi/Auth/JwtService.fs
module DemoBeFsgi.Auth.JwtService

open System
open System.IdentityModel.Tokens.Jwt
open System.Security.Claims
open System.Text
open Microsoft.IdentityModel.Tokens

let private getSecret () =
    let s = Environment.GetEnvironmentVariable("APP_JWT_SECRET")

    if String.IsNullOrEmpty s then
        "dev-jwt-secret-at-least-32-characters-long-for-hmac"
    else
        s

let private getSigningKey () =
    let key = Encoding.UTF8.GetBytes(getSecret ())
    SymmetricSecurityKey(key)

let generateAccessToken (userId: Guid) (username: string) (email: string) (role: string) : string =
    let now = DateTime.UtcNow
    let jti = Guid.NewGuid().ToString()

    let claims =
        [| Claim(JwtRegisteredClaimNames.Sub, userId.ToString())
           Claim(JwtRegisteredClaimNames.Jti, jti)
           Claim("username", username)
           Claim("email", email)
           Claim("role", role) |]

    let signingCreds =
        SigningCredentials(getSigningKey (), SecurityAlgorithms.HmacSha256)

    let token =
        JwtSecurityToken(
            issuer = "demo-be-fsharp-giraffe",
            audience = "demo-be-fsharp-giraffe",
            claims = claims,
            notBefore = now,
            expires = now.AddMinutes(15.0),
            signingCredentials = signingCreds
        )

    JwtSecurityTokenHandler().WriteToken(token)

let generateRefreshToken (userId: Guid) : string =
    let now = DateTime.UtcNow
    let jti = Guid.NewGuid().ToString()

    let claims =
        [| Claim(JwtRegisteredClaimNames.Sub, userId.ToString())
           Claim(JwtRegisteredClaimNames.Jti, jti)
           Claim("token_type", "refresh") |]

    let signingCreds =
        SigningCredentials(getSigningKey (), SecurityAlgorithms.HmacSha256)

    let token =
        JwtSecurityToken(
            issuer = "demo-be-fsharp-giraffe",
            audience = "demo-be-fsharp-giraffe",
            claims = claims,
            notBefore = now,
            expires = now.AddDays(7.0),
            signingCredentials = signingCreds
        )

    JwtSecurityTokenHandler().WriteToken(token)

let validateToken (token: string) : ClaimsPrincipal option =
    let handler = JwtSecurityTokenHandler()
    let key = getSigningKey ()

    let validationParams =
        TokenValidationParameters(
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = key,
            ValidateIssuer = true,
            ValidIssuer = "demo-be-fsharp-giraffe",
            ValidateAudience = true,
            ValidAudience = "demo-be-fsharp-giraffe",
            ValidateLifetime = true,
            ClockSkew = TimeSpan.Zero
        )

    try
        let mutable securityToken: SecurityToken = null
        let principal = handler.ValidateToken(token, validationParams, &securityToken)
        Some principal
    with _ ->
        None

let getTokenJti (token: string) : string option =
    try
        let handler = JwtSecurityTokenHandler()
        let jwt = handler.ReadJwtToken(token)

        let jtiClaim =
            jwt.Claims |> Seq.tryFind (fun c -> c.Type = JwtRegisteredClaimNames.Jti)

        jtiClaim |> Option.map (fun c -> c.Value)
    with _ ->
        None

let getTokenExpiry (token: string) : DateTime option =
    try
        let handler = JwtSecurityTokenHandler()
        let jwt = handler.ReadJwtToken(token)
        Some jwt.ValidTo
    with _ ->
        None

let getJwks () =
    let key = getSigningKey ()
    let keyId = "demo-be-fsharp-giraffe-key-1"

    let jwk =
        {| kty = "oct"
           kid = keyId
           alg = "HS256"
           use_ = "sig"
           k = Convert.ToBase64String(key.Key) |}

    {| keys =
        [| {| kty = jwk.kty
              kid = jwk.kid
              alg = jwk.alg
              ``use`` = jwk.use_ |} |] |}
